jina/executors/indexers/vector.py

```diff
@@ -70,15 +72,36 @@ def __init__(self,
 
     def post_init(self):
         super().post_init()
-        self.valid_indices = np.array([], dtype=bool)
 
     def _delete_invalid_indices(self):
-        """Here rebuild the index with only valid entries"""
-        pass
+        # problem is data is not flushed there yet.
+        # self.write_handler.close()
+        # self.handler_mutex = False
+        valid = self.valid_indices[self.valid_indices == True]  # noqa
+        if len(valid) != len(self.valid_indices):
 
-    def __getstate__(self):
+            deleted_keys = len(self.valid_indices[self.valid_indices == False])
+            tmp_ndarray = np.memmap(self.index_abspath, dtype=self.dtype, mode='r',
+                                 shape=(self.size + deleted_keys, self.num_dim))
+            open(self.index_abspath+'tmp', 'wb')
+            valid_ndarray = np.memmap(self.index_abspath+'tmp', dtype=self.dtype, mode='r+',
+                                 shape=(len(valid), self.num_dim))
+            valid_ndarray[:] = tmp_ndarray[self.valid_indices]
+            valid_ndarray.flush()
+            del tmp_ndarray
+            os.remove(self.index_abspath)
+            os.rename(self.index_abspath+'tmp', self.index_abspath)
+
+            valid_key_bytes = np.frombuffer(self.key_bytes, dtype=(np.str_, self.key_length))[self.valid_indices].tobytes()
+            self.key_bytes = valid_key_bytes
+            self._size = len(valid)
+            self.valid_indices = valid
+            # self._int2ext_id = self._int2ext_id[valid]
+
+    def __setstate__(self, d):
```

Comment:
why do we need to override?

Reply:
I see why, but isnit it better to delete on store? why would we store data we actually do not want?